# Import Library

In [1]:
import numpy as np 

import pandas as pd 
#code untuk membuat kolom tabel dapat ditampilkan semua 
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',99)

import matplotlib.pyplot as plt 
import seaborn as sns 
sns.set()

In [2]:
df = pd.read_csv('loan_data_2007_2014.csv', index_col=0)
df.head()

C:\Users\ASUS\AppData\Local\Temp\ipykernel_5484\1883642690.py:1: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('loan_data_2007_2014.csv', index_col=0)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

   int_rate  installment grade sub_grade                 emp_title emp_length  \
0     10.65       162.87     B        B2                       NaN  10+ years   
1     15.27        59.83     C        C4                     Ryder   < 1 year   
2     15.96        84.33     C        C5                       NaN  10+ years   
3     13.49       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4     12.69        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://www.lendingclub.com/browse/loanDetail....   
1          n  https://www.lendingclub.com/browse/loanDetail....   
2          n  https://www.lendingclub.com/browse/loanDetail....   
3          n  https://www.lendingclub.com/browse/loanDetail....   
4          n  https://www.lendingclub.com/browse/loanDetail....   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65          0.0   
1                  bike    309xx         GA   1.00          0.0   
2  real estate business    606xx         IL   8.72          0.0   
3              personel    917xx         CA  20.00          0.0   
4              Personal    972xx         OR  17.94          0.0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85             1.0                     NaN   
1           Apr-99             5.0                     NaN   
2           Nov-01             2.0                     NaN   
3           Feb-96             1.0                    35.0   
4           Jan-96             0.0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
0                     NaN       3.0      0.0      13648        83.7   
1                     NaN       3.0      0.0       1687         9.4   
2                     NaN       2.0      0.0       2956        98.5   
3                     NaN      10.0      0.0       5598        21.0   
4                     NaN      15.0      0.0      27783        53.9   

   total_acc initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0        9.0                   f        0.0            0.0   5861.071414   
1        4.0                   f        0.0            0.0   1008.710000   
2       10.0                   f        0.0            0.0   3003.653644   
3       37.0                   f        0.0            0.0  12226.302210   
4       38.0                   f      766.9          766.9   3242.170000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5831.78          5000.00        

# Exploring Data 

In [3]:
df.shape

(466285, 74)

In [4]:
df.sample()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
418088  13266522   15308772      13275        13275          13275.0   

              term  int_rate  installment grade sub_grade         emp_title  \
418088   36 months      7.62       413.67     A        A3  School principal   

       emp_length home_ownership  annual_inc verification_status issue_d  \
418088  10+ years            OWN    103000.0        Not Verified  Mar-14   

       loan_status pymnt_plan  \
418088     Current          n   

                                                      url desc  \
418088  https://www.lendingclub.com/browse/loanDetail....  NaN   

               purpose           title zip_code addr_state   dti  delinq_2yrs  \
418088  major_purchase  Major purchase    480xx         MI  1.47          0.0   

       earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
418088           Jun-94             0.0                     NaN   

        mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
418088                     NaN       6.0      0.0      25925        49.6   

        total_acc initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
418088       11.0                   w    5900.75        5900.75      8687.07   

        total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
418088          8687.07          7374.25        1312.82                 0.0   

        recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
418088         0.0                      0.0       Dec-15           413.67   

       next_pymnt_d last_credit_pull_d  collections_12_mths_ex_med  \
418088       Jan-16             Jan-16                         0.0   

        mths_since_last_major_derog  policy_code application_type  \
418088                          NaN            1       INDIVIDUAL   

        annual_inc_joint  dti_joint  verification_status_joint  \
418088               NaN        NaN                        NaN   

        acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_il_6m  \
418088             0.0           0.0     134816.0          NaN         NaN   

        open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  \
418088          NaN          NaN                 NaN           NaN      NaN   

        open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  \
418088          NaN          NaN         NaN       NaN           52300.0   

        inq_fi  total_cu_tl  inq_last_12m  
418088     NaN          NaN           NaN

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 466285 entries, 0 to 466284
Data columns (total 74 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           466285 non-null  int64  
 1   member_id                    466285 non-null  int64  
 2   loan_amnt                    466285 non-null  int64  
 3   funded_amnt                  466285 non-null  int64  
 4   funded_amnt_inv              466285 non-null  float64
 5   term                         466285 non-null  object 
 6   int_rate                     466285 non-null  float64
 7   installment                  466285 non-null  float64
 8   grade                        466285 non-null  object 
 9   sub_grade                    466285 non-null  object 
 10  emp_title                    438697 non-null  object 
 11  emp_length                   445277 non-null  object 
 12  home_ownership               466285 non-null  object 
 13 

Pengamatan
1. Data terdiri dari 466285 baris
2. Tampak beberapa kolom masih memiliki null/missing values (Non-Null Count < jumlah baris)
3. Tipe kolom emp_length, issue_d , earliest_cr_line, last_pymnt_d, next_pymnt_d, last_credit_pull_d memiliki tipe yang salah (seharusnya dalam bentuk int), sedangkan kolom lainnya sepertinya tidak ada issue yang mencolok pada tipe data (sudah sesuai)
4. Ada beberapa fitur kolom tidak berguna karena memiliki nilai kosong semua (Null) 

In [6]:
df.id.nunique()

466285

In [7]:
df.member_id.nunique()

466285

Terlihat bahwa tidak ada `id` atau `member_id` yang duplikat, artinya setiap baris sudah mewakili satu individu.

Selanjutnya, pembuangan fitur-fitur yang tidak berguna dilakukan. Contohnya seperti fitur yang merupakan id unik, berupa free text, nilainya kosong semua (NULL), dsb.

# Data Cleaning 

In [8]:
df.isna().sum()

id                                  0
member_id                           0
loan_amnt                           0
funded_amnt                         0
funded_amnt_inv                     0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_title                       27588
emp_length                      21008
home_ownership                      0
annual_inc                          4
verification_status                 0
issue_d                             0
loan_status                         0
pymnt_plan                          0
url                                 0
desc                           340302
purpose                             0
title                              20
zip_code                            0
addr_state                          0
dti                                 0
delinq_2yrs                        29
earliest_cr_

#### Jika ada kolom dengan data kosong yang sangat banyak, maka kolom tersebut bisa dihapus aja

In [9]:
# kolom all null 
cols_to_drop= ['inq_last_12m','total_cu_tl','inq_fi','open_acc_6m','open_il_6m','open_il_12m','open_il_24m','mths_since_rcnt_il','total_bal_il','il_util','open_rv_12m','open_rv_24m','max_bal_bc','all_util','annual_inc_joint','dti_joint', 'verification_status_joint'
              # drop kolom free text (data  tidak berguna)
, 'url','desc']

In [10]:
df_1 = df.drop(cols_to_drop, axis=1)

In [11]:
df_1.sample(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
34321     454473     562832       7000         7000           7000.0   
401959  14789207   16861633       3375         3375           3375.0   
183966   1904826    2216993      10000        10000          10000.0   
45230   10065862   11917935       2400         2400           2400.0   
277949  31217306   33790535       5000         5000           5000.0   

              term  int_rate  installment grade sub_grade  \
34321    36 months      7.40       217.42     A        A2   
401959   36 months     16.59       119.65     D        D1   
183966   36 months     13.11       337.47     B        B4   
45230    36 months     13.98        82.01     C        C1   
277949   36 months      6.49       153.23     A        A2   

                     emp_title emp_length home_ownership  annual_inc  \
34321             City of Mesa   < 1 year       MORTGAGE     40224.0   
401959               Head Cook  10+ years       MORTGAGE     39000.0   
183966  kokosing construction     2 years           RENT     60000.0   
45230          Program Manager  10+ years       MORTGAGE     75000.0   
277949                 Manager    5 years            OWN     45000.0   

       verification_status issue_d  loan_status pymnt_plan  \
34321         Not Verified  Nov-09   Fully Paid          n   
401959            Verified  Apr-14      Current          n   
183966        Not Verified  Nov-12  Charged Off          n   
45230      Source Verified  Dec-13      Current          n   
277949        Not Verified  Oct-14      Current          n   

                   purpose                    title zip_code addr_state  \
34321          credit_card             sikofmyrates    852xx         AZ   
401959  debt_consolidation       Debt consolidation    641xx         MO   
183966               other                    Other    152xx         PA   
45230              medical            Medical bills    786xx         TX   
277949         credit_card  Credit card refinancing    441xx         OH   

          dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
34321    8.89          0.0           Dec-84             2.0   
401959  17.88          0.0           Nov-93             1.0   
183966  17.72          0.0           Mar-94             0.0   
45230    9.25          0.0           Sep-00             0.0   
277949  19.86          0.0           May-03             1.0   

        mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
34321                      NaN                     NaN      14.0      0.0   
401959                     NaN                    58.0       5.0      1.0   
183966                    54.0                     NaN       6.0      0.0   
45230                     34.0                     NaN       8.0      0.0   
277949                    63.0                     NaN      13.0      0.0   

        revol_bal  revol_util  total_acc initial_list_status  out_prncp  \
34321       14014        14.1       36.0                   f       0.00   
401959       2311        92.0        9.0                   w    1706.58   
183966      17820        68.0       20.0                   w       0.00   
45230           0         0.0       23.0                   f     913.51   
277949      21300        58.5       31.0                   w    3169.90   

        out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
34321            0.00      7800.19          7800.19          7000.00   
401959        1706.58      2393.00          2393.00          1668.42   
183966           0.00      3034.98          3034.98          2144.48   
45230          913.51      1967.78          1967.78          1486.49   
277949        3169.90      2152.35          2152.35          1830.10   

        total_rec_int  total_rec_late_fee  recoveries  \
34321          800.19                 0.0         0.0   
401959         724.58                 0.0         0.0   
183966         890.50                 0.0         0.0   
452

In [13]:
df_1.isna().sum()

id                                  0
member_id                           0
loan_amnt                           0
funded_amnt                         0
funded_amnt_inv                     0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_title                       27588
emp_length                      21008
home_ownership                      0
annual_inc                          4
verification_status                 0
issue_d                             0
loan_status                         0
pymnt_plan                          0
purpose                             0
title                              20
zip_code                            0
addr_state                          0
dti                                 0
delinq_2yrs                        29
earliest_cr_line                   29
inq_last_6mths                     29
mths_since_l